### PR TITLE
Task/gps and location sharing/gps fetcher

### PR DESCRIPTION
**Pull request by BarAgam**
GPS fetcher class that wraps android's location services.
Includes Unit-tests.
MainActivity holds full usage example that displays location with a toast

### CODE REVIEW

**Comment by review-ninja** [comment]
<a href="https://app.review.ninja/teamagam/GimelGimel/pull/10" target="_blank">ReviewNinja

File: app/src/main/java/com/teamagam/gimelgimel/app/control/sensors/LocationFetcher.java [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+public class LocationFetcher {
+
+    @StringDef({
+            ProviderType.LOCATION_PROVIDER_GPS,
+            ProviderType.LOCATION_PROVIDER_NETWORK,
+            ProviderType.LOCATION_PROVIDER_PASSIVE
+    })
+    @Retention(RetentionPolicy.SOURCE)
+    public @interface ProviderType {
+        String LOCATION_PROVIDER_GPS = LocationManager.GPS_PROVIDER;
```
Comment (teamagam1) [inline app/src/main/java/com/teamagam/gimelgimel/app/control/sensors/LocationFetcher.java:35]: do we really need it?
why don't just use LocationManager.XXX_PROVIDER

File: app/src/main/java/com/teamagam/gimelgimel/app/control/sensors/LocationFetcher.java [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+        String LOCATION_PROVIDER_NETWORK = LocationManager.NETWORK_PROVIDER;
+        String LOCATION_PROVIDER_PASSIVE = LocationManager.PASSIVE_PROVIDER;
+    }
+
+
+    private static void checkForLocationPermission(Context context) {
```
Comment (teamagam1) [inline app/src/main/java/com/teamagam/gimelgimel/app/control/sensors/LocationFetcher.java:41]: consider using context only for checking permission once in the constructor and do not hold field with pointer to the context.

File: app/src/main/java/com/teamagam/gimelgimel/app/control/sensors/LocationFetcher.java [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+
+    /**
+     * Registers fetcher for location updates
+     *
+     * @param minSamplingFrequencyMs         - minimum time between location samples,  in milliseconds
+     * @param minDistanceDeltaSamplingMeters - minimum distance between location samples, in meters
+     */
+    public void registerForUpdates(long minSamplingFrequencyMs,
+                                   float minDistanceDeltaSamplingMeters) {
+        if (minSamplingFrequencyMs < 0) {
```
Comment (teamagam1) [inline app/src/main/java/com/teamagam/gimelgimel/app/control/sensors/LocationFetcher.java:105]: 0 or very small number is OK? how small we allow?

File: app/src/main/java/com/teamagam/gimelgimel/app/control/sensors/LocationFetcher.java [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+            throw new IllegalArgumentException("minSamplingFrequencyMs cannot be negative");
+        }
+
+        if (minDistanceDeltaSamplingMeters < 0) {
```
Comment (teamagam1) [inline app/src/main/java/com/teamagam/gimelgimel/app/control/sensors/LocationFetcher.java:109]: same

File: app/src/main/java/com/teamagam/gimelgimel/app/control/sensors/LocationFetcher.java [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+            throw new IllegalArgumentException(
+                    "minDistanceDeltaSamplingMeters cannot be a negative number");
+        }
+
+        if (mIsRegistered) {
```
Comment (teamagam1) [inline app/src/main/java/com/teamagam/gimelgimel/app/control/sensors/LocationFetcher.java:114]: I'm not sure about it, if someone else also want location, why does he need to know about someone else? maybe add isRegistered method.

File: app/src/main/java/com/teamagam/gimelgimel/app/control/sensors/LocationFetcher.java [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.util.ArrayList;
+import java.util.Collection;
+
+/**
+ * Handles location fetching against the system's sensors
+ */
+public class LocationFetcher {
```
Comment (teamagam1) [inline app/src/main/java/com/teamagam/gimelgimel/app/control/sensors/LocationFetcher.java:24]: does it going to be a singelton?

File: app/src/main/java/com/teamagam/gimelgimel/app/control/sensors/LocationFetcher.java [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+        @Override
+        public void onLocationChanged(Location location) {
+            LocationSample locationSample = new LocationSample(location);
+            mLocationFetcherListener.onNewLocationSample(locationSample);
+        }
+
+
+        @Override
+        public void onStatusChanged(String provider, int status, Bundle extras) {
+            if (status == android.location.LocationProvider.AVAILABLE) {
```
Comment (teamagam1) [inline app/src/main/java/com/teamagam/gimelgimel/app/control/sensors/LocationFetcher.java:176]: what 'status' holds for? why do you need to go to android.location.LocationProvider.AVAILABLE?

File: app/src/main/java/com/teamagam/gimelgimel/app/control/sensors/LocationFetcher.java [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+    }
+
+    /**
+     * Encapsulates incoming updates handling from {@link LocationFetcher}'s underlying {@link LocationManager}
+     */
+    private class MyLocationListener implements LocationListener {
+
+        @Override
+        public void onLocationChanged(Location location) {
+            LocationSample locationSample = new LocationSample(location);
```
Comment (teamagam1) [inline app/src/main/java/com/teamagam/gimelgimel/app/control/sensors/LocationFetcher.java:168]: add Log !fix

File: app/src/main/java/com/teamagam/gimelgimel/app/network/rest/RestAPI.java [hunk 9]
```diff
@@ -41,7 +43,13 @@ public GGMessagingAPI getMessagingAPI() {
 
         private static GGMessagingAPI initializeMessagingAPI() {
             //http logger for debugging
-            HttpLoggingInterceptor interceptor = new HttpLoggingInterceptor();
+            HttpLoggingInterceptor interceptor = new HttpLoggingInterceptor(
+                    new HttpLoggingInterceptor.Logger() {
+                        @Override
+                        public void log(String message) {
+                            Log.v("GGMessagingHttpClient", message);
```
Comment (teamagam1) [inline app/src/main/java/com/teamagam/gimelgimel/app/network/rest/RestAPI.java:50]: all this code is for changing the log to verbose. right?

File: app/src/main/java/com/teamagam/gimelgimel/app/control/sensors/LocationFetcher.java [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+        mProviders = new ArrayList<>();
+        mIsRegistered = false;
+    }
+
+    /**
+     * Adds provider to be used when registering the fetcher
+     *
+     * @param locationProvider - type of provider to use for location updates
+     */
+    public void addProvider(@ProviderType String locationProvider) {
```
Comment (teamagam1) [inline app/src/main/java/com/teamagam/gimelgimel/app/control/sensors/LocationFetcher.java:84]: don't we want to add all providers available always?

File: app/src/main/java/com/teamagam/gimelgimel/app/control/sensors/LocationFetcher.java (hunk 6 above)
Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/control/sensors/LocationFetcher.java:24]: No

File: app/src/main/java/com/teamagam/gimelgimel/app/control/sensors/LocationFetcher.java (hunk 1 above)
Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/control/sensors/LocationFetcher.java:35]: Used to to enforce valid values at compile time.

File: app/src/main/java/com/teamagam/gimelgimel/app/control/sensors/LocationFetcher.java (hunk 2 above)
Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/control/sensors/LocationFetcher.java:41]: LINT advises to check for the permission right before updates requests.

File: app/src/main/java/com/teamagam/gimelgimel/app/control/sensors/LocationFetcher.java (hunk 10 above)
Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/control/sensors/LocationFetcher.java:84]: Its up to the user of this class

File: app/src/main/java/com/teamagam/gimelgimel/app/control/sensors/LocationFetcher.java (hunk 3 above)
Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/control/sensors/LocationFetcher.java:105]: Anything greater than or equals to 0 is OK.
0 means "give me the updates as frequent as possible"
The actual update frequency depends on the device and the state its at

File: app/src/main/java/com/teamagam/gimelgimel/app/control/sensors/LocationFetcher.java (hunk 5 above)
Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/control/sensors/LocationFetcher.java:114]: It should use another LocationFetcher.
It keeps a fetcher from getting new type of sources after registration.
I've also added a conditional at addProvider to help with that.

File: app/src/main/java/com/teamagam/gimelgimel/app/control/sensors/LocationFetcher.java (hunk 7 above)
Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/control/sensors/LocationFetcher.java:176]: android.location.LocationProvider hold a few static ints that represent the status result domain.
Available is the only one that means the provider is enabled.
Just to make it clear, it's the Android LocationListener that im wrapping. A user of LocationFetcher isn't aware of this status.

File: app/src/main/java/com/teamagam/gimelgimel/app/network/rest/RestAPI.java (hunk 9 above)
Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/network/rest/RestAPI.java:50]: yes

File: app/src/main/java/com/teamagam/gimelgimel/app/control/sensors/LocationFetcher.java (hunk 8 above)
Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/control/sensors/LocationFetcher.java:168]: !fixed

**Pull request merged**